kvbc/src/ReplicaImp.cpp

```diff
@@ -319,13 +310,42 @@ bool ReplicaImp::putBlock(const uint64_t blockId, const char *block_data, const
   return true;
 }
 
+uint64_t ReplicaImp::getLastReachableBlockNum() const {
+  if (replicaConfig_.isReadOnly) {
+    return m_bcDbAdapter->getLastReachableBlockId();
+  }
+  return m_kvBlockchain->getLastReachableBlockId();
+}
+
+uint64_t ReplicaImp::getLastBlockNum() const {
+  if (replicaConfig_.isReadOnly) {
+    return m_bcDbAdapter->getLatestBlockId();
+  }
+  const auto last = m_kvBlockchain->getLastStatetransferBlockId();
+  if (last) {
+    return *last;
+  }
+  return m_kvBlockchain->getLastReachableBlockId();
+}
+
 RawBlock ReplicaImp::getBlockInternal(BlockId blockId) const { return m_bcDbAdapter->getRawBlock(blockId); }
 
 /*
  * This method assumes that *outBlock is big enough to hold block content
  * The caller is the owner of the memory
  */
 bool ReplicaImp::getBlock(uint64_t blockId, char *outBlock, uint32_t *outBlockSize) {
+  if (replicaConfig_.isReadOnly) {
+    return getBlockFromObjectStore(blockId, outBlock, outBlockSize);
+  }
+  const auto rawBlock = m_kvBlockchain->getRawBlock(blockId);
+  const auto &ser = categorization::RawBlock::serialize(rawBlock);
+  *outBlockSize = ser.size();
+  std::memcpy(outBlock, ser.data(), *outBlockSize);
```

Comment:
😢

Reply:
Yeah, the API of this method is problematic. Leaving for a future commit.